client/js/Core/componentsandbox.js

```diff
@@ -260,289 +274,410 @@ SCWeb.core.ComponentSandbox.prototype.createViewersForScStructs = function (cont
+                        levelStyles, visitedElements, relationElements);
+                    if (Object.keys(newElements).length) newElementsArr.push(newElements);
                 }
-                await searchAllLevelEdges(newElementsArr, levelScales, level + 1, visitedElements);
+                await searchAllLevelEdges(newElementsArr, allLevelStyles, level + 1, visitedElements, relationElements);
             }
         };
 
-        let searchLevelEdges = async function (mainElem, scale, visitedElements) {
-            let outgoingLevelNodesWithRelation = await searchLevelEdgesByDirection(mainElem, scale, visitedElements, false, true);
-            let outgoingLevelNodesNotWithRelation = await searchLevelEdgesByDirection(mainElem, scale, visitedElements, false, false);
-            let incomingLevelNodesNotWithRelation = await searchLevelEdgesByDirection(mainElem, scale, visitedElements, true, false);
-            let incomingLevelNodesWithRelation = await searchLevelEdgesByDirection(mainElem, scale, visitedElements, true, true);
-            return [...incomingLevelNodesWithRelation, ...incomingLevelNodesNotWithRelation, ...outgoingLevelNodesWithRelation, ...outgoingLevelNodesNotWithRelation];
+        const searchLevelEdges = async function (mainElement, mainElementType, prevLevelStyles, levelStyles, visitedElements, relationElements) {
+            let levelNodes = {};
+            await searchLevelEdgesByDirection(
+                mainElement, mainElementType, prevLevelStyles, levelStyles,
+                visitedElements, levelNodes, relationElements,
+                false, true, false);
+            await searchLevelEdgesByDirection(
+                mainElement, mainElementType, prevLevelStyles, levelStyles,
+                visitedElements, levelNodes, relationElements,
+                false, false, false);
+            await searchLevelEdgesByDirection(
+                mainElement, mainElementType, prevLevelStyles, levelStyles,
+                visitedElements, levelNodes, relationElements,
+                true, false, false);
+            await searchLevelEdgesByDirection(
+                mainElement, mainElementType, prevLevelStyles, levelStyles,
+                visitedElements, levelNodes, relationElements,
+                true, true, true);
+            return levelNodes;
         };
 
-        let searchLevelEdgesByDirection = async function (mainElem, scale, visitedElements, incomingEdge, withRelation) {
-            let levelNodes = [];
-
-            let scTemplateSearchEdgeElements = new sc.ScTemplate();
+        const searchLevelEdgesByDirection = async function (
+            mainElement, mainElementType, prevLevelStyles, levelStyles,
+            visitedElements, levelNodes, relationElements,
+            isIncomingEdge, withRelation, withEdgeToEdge) {
 
-            if (edgeToEdge) {
+            if (withEdgeToEdge) {
```

Comment:
this if looks like it belongs to its own separate method. it took me a lot of time to understand why there are 3 boolean flags but this method is called 4 times instead of 8 times